lib/cache/sqlite-cache-store.js

```diff
@@ -0,0 +1,399 @@
+     */
+    const result = {
+      body: value.body ? parseBufferArray(JSON.parse(value.body)) : null,
+      statusCode: value.statusCode,
+      statusMessage: value.statusMessage,
+      rawHeaders: parseBufferArray(JSON.parse(value.rawHeaders)),
+      cachedAt: value.cachedAt,
+      staleAt: value.staleAt,
+      deleteAt: value.deleteAt
+    }
+
+    return result
+  }
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.CacheKey} key
+   * @param {import('../../types/cache-interceptor.d.ts').default.CacheValue} value
+   * @returns {Writable | undefined}
+   */
+  createWriteStream (key, value) {
+    if (typeof key !== 'object') {
+      throw new TypeError(`expected key to be object, got ${typeof key}`)
+    }
+
+    if (typeof value !== 'object') {
+      throw new TypeError(`expected value to be object, got ${typeof value}`)
+    }
+
+    const url = this.#makeValueUrl(key)
+    let currentSize = 0
+    /**
+     * @type {Buffer[] | null}
+     */
+    let body = key.method !== 'HEAD' ? [] : null
+    const maxSize = this.#maxSize
+    const findValue = this.#findValue.bind(this)
+    const updateValueQuery = this.#updateValueQuery
+    const insertValueQuery = this.#insertValueQuery
+
+    const writable = new Writable({
+      write (chunk, encoding, callback) {
+        if (typeof chunk === 'string') {
+          chunk = Buffer.from(chunk, encoding)
+        }
+
+        currentSize += chunk.byteLength
+
+        if (body) {
+          if (currentSize >= maxSize) {
+            body = null
+            this.end()
+            return callback()
+          }
+
+          body.push(chunk)
+        }
+
+        callback()
+      },
+      final (callback) {
+        if (body === null) {
+          return callback()
+        }
+
+        /**
+         * @type {SqliteStoreValue | undefined}
+         */
+        const existingValue = findValue(key, true)
+        if (existingValue) {
+          // Updating an existing response, let's delete it
+          updateValueQuery.run(
+            JSON.stringify(stringifyBufferArray(body)),
+            value.deleteAt,
+            value.statusCode,
+            value.statusMessage,
+            JSON.stringify(stringifyBufferArray(value.rawHeaders)),
+            value.cachedAt,
+            value.staleAt,
+            value.deleteAt,
+            existingValue.id
+          )
+        } else {
+          // New response, let's insert it
+          insertValueQuery.run(
+            url,
+            key.method,
+            JSON.stringify(stringifyBufferArray(body)),
+            value.deleteAt,
+            value.statusCode,
+            value.statusMessage,
+            JSON.stringify(stringifyBufferArray(value.rawHeaders)),
+            value.vary ? JSON.stringify(value.vary) : null,
+            value.cachedAt,
+            value.staleAt,
+            value.deleteAt
+          )
+        }
```

Comment:
Can't you just always just overwrite? Does it update and insert need to be separate ops?

Reply:
I think so but I'm not sure how messy it would be compared to just having two queries

Reply:
Does update fail if the value doesn't exist? Doesn't sql have an insert or update command?

Reply:
> Does update fail if the value doesn't exist

It fails in the sense that it noops.

> Doesn't sql have an insert or update command?

I'm not sure how well insert or update would work here based on how we're checking if there's a already cached value though. The only reliable and fast way we have to compare two cached values is with their ids. Otherwise, we need to compare their url, method, and vary headers to see if they were the same. The url and method comparisons are simple and can be done in sql, but comparing the vary headers is more difficult since it's stringified json. We can't compare it in sql since we don't know if the properties are in the same order, meaning we need to parse then compare (like what `#findValues` does).

I was experimenting with hashing the vary headers (sort properties alphabetically -> stringify -> crypto.createHash), however, it was a lot more complicated than just having two separate queries. It also would've only benefited writes, for reads we would still need to fetch each response and compare the vary headers like we already do